### PR TITLE
Feature/time slot

### PR DESCRIPTION
**Pull request by MrZyr0**

### CODE REVIEW

File: src/models/time-slot.model.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+@Schema()
+export class TimeSlot {
+  _id: Types.ObjectId;
+
+  @Prop({ required: true, unique: true })
+  groupId: string;
+
+  @Prop({ required: true })
+  name: string;
```
Comment (MaxenceMottard) [inline src/models/time-slot.model.ts:14]: really necessary ?

File: src/modules/app.module.ts [hunk 2]
```diff
@@ -23,13 +25,15 @@ import { MailService } from '../services/mail.service';
...
     MailService,
+    TimeSlotService,
   ],
   controllers: [
     AppController,
     AuthenticationController,
     OrganizationController,
     GroupController,
     FileController,
+    TimeSlotController,
```
Comment (MaxenceMottard) [inline src/modules/app.module.ts:36]:
```suggestion
    TimeSlotController,
    FileController,
```

Comment (MaxenceMottard) [reply src/modules/app.module.ts:36]: Just for Swagger, I prefer to leave the FileController last 😅

File: src/services/time-slot.service.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      { ...set },
+    );
+
+  deleteOneGroupTimeSlotById = async (timeSlotId: string): Promise<boolean> =>
+    this.timeSlotRepository.deleteOnyBy({ timeSlotId });
+
+  getUserTimeSlots = async (userId: string): Promise<TimeSlot[]> =>
+    this.timeSlotRepository.findManyBy({
+      userId,
+    });
```
Comment (MaxenceMottard) [inline src/services/time-slot.service.ts:47]:
```suggestion
    this.timeSlotRepository.findManyBy({ userId });
```

File: src/services/time-slot.service.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+  getAllGroupTimeSlots = async (groupId: string): Promise<TimeSlot[]> =>
+    this.timeSlotRepository.findManyBy({
+      groupId,
+    });
+
+  getOneGroupTimeSlotById = async (timeSlotId: string): Promise<TimeSlot> =>
+    this.timeSlotRepository.findOneBy({
+      _id: timeSlotId,
+    });
```
Comment (MaxenceMottard) [inline src/services/time-slot.service.ts:24]:
```suggestion
    this.timeSlotRepository.findOneBy({  _id: timeSlotId });
```

File: src/services/time-slot.service.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  insertTimeSlot = async (
+    groupId: string,
+    parameters: TimeSlotDTO,
+  ): Promise<TimeSlot> =>
+    this.timeSlotRepository.insert({ groupId, ...parameters });
+
+  getAllGroupTimeSlots = async (groupId: string): Promise<TimeSlot[]> =>
+    this.timeSlotRepository.findManyBy({
+      groupId,
+    });
```
Comment (MaxenceMottard) [inline src/services/time-slot.service.ts:19]:
```suggestion
    this.timeSlotRepository.findManyBy({ groupId });
```

File: src/controllers/time-slot.controller.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    @Body() body: TimeSlotDTO,
+  ) {
+    return this.timeSlotService.updateOneGroupTimeSlotById(
+      {
+        groupId,
+        timeSlotId,
+      },
+      {
+        ...body,
+      },
```
Comment (MaxenceMottard) [inline src/controllers/time-slot.controller.ts:60]:
```suggestion
      { groupId, timeSlotId },
      {  ...body },
```

Comment (MaxenceMottard) [reply src/controllers/time-slot.controller.ts:60]: Just for Swagger, I prefer to leave the FileController last 😅

File: src/models/time-slot.model.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import { Document, Types } from 'mongoose';
+
+export type TimeSlotDocument = TimeSlot & Document;
+
+@Schema()
+export class TimeSlot {
+  _id: Types.ObjectId;
+
+  @Prop({ required: true, unique: true })
+  groupId: string;
```
Comment (MaxenceMottard) [inline src/models/time-slot.model.ts:11]:
```suggestion
  @Prop({ required: true, type: Types.ObjectId, ref: Group.name })
  @Prop({ required: true })
  group: Group;
```

Comment (MaxenceMottard) [reply src/models/time-slot.model.ts:11]: I'm not sure about unique property

**Review by MaxenceMottard — CHANGES_REQUESTED** [review]

File: src/models/time-slot.model.ts (hunk 1 above)
Comment (MrZyr0) [reply src/models/time-slot.model.ts:14]: Nope 😅

File: src/controllers/time-slot.controller.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    @Body() body: PatchingTimeSlotDTO,
+  ) {
+    return this.timeSlotService.updateOneGroupTimeSlotById(
+      { groupId, timeSlotId },
+      { ...body },
+    );
+  }
+
+  @Delete('/:timeSlotId')
+  @UseGuards(JWTGuard, OrganizationAdminGuard)
```
Comment (MaxenceMottard) [inline src/controllers/time-slot.controller.ts:64]: OrganizationAdminGuard doesn't work without `:organizationId`

Comment (MaxenceMottard) [reply src/controllers/time-slot.controller.ts:64]: You should create a guard for this and `time-slot.controller.ts:42`

File: src/services/time-slot.service.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    },
+    set: PatchingTimeSlotDTO,
+  ): Promise<boolean> =>
+    this.timeSlotRepository.updateOneBy(
+      {
+        group: new mongoose.Types.ObjectId(conditions.groupId),
+        timeSlotId: conditions.timeSlotId,
+      },
+      {
+        ...set,
```
Comment (MaxenceMottard) [inline src/services/time-slot.service.ts:47]: bad idea, `PatchingTimeSlotDTO` contains a `groupId` property that is not in TimeSlot model

File: src/services/time-slot.service.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    conditions: {
+      groupId: string;
+      timeSlotId: string;
+    },
+    set: PatchingTimeSlotDTO,
+  ): Promise<boolean> =>
+    this.timeSlotRepository.updateOneBy(
+      {
+        group: new mongoose.Types.ObjectId(conditions.groupId),
+        timeSlotId: conditions.timeSlotId,
```
Comment (MaxenceMottard) [inline src/services/time-slot.service.ts:44]:
```suggestion
        _id: conditions.timeSlotId,
```

File: src/services/time-slot.service.ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+      },
+      {
+        ...set,
+        group: set.groupdId
+          ? new mongoose.Types.ObjectId(set.groupdId)
+          : undefined,
```
Comment (MaxenceMottard) [inline src/services/time-slot.service.ts:50]:
```suggestion
```

Comment (MaxenceMottard) [reply src/services/time-slot.service.ts:50]: it should not be possible to modify the group of a time slot

File: src/services/time-slot.service.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  getUserTimeSlots = async (user: User) => {
+    const userGroups = await this.groupRepository.findManyBy({
+      users: {
+        $in: user._id,
+      },
+    });
+
+    const allTimeSlots = await Promise.all(
+      userGroups.map((group) =>
+        this.timeSlotRepository.findManyBy({ groupeId: group._id }),
```
Comment (MaxenceMottard) [inline src/services/time-slot.service.ts:66]:
```suggestion
        this.getAllGroupTimeSlots(group._id),
```

File: src/services/time-slot.service.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    });
+
+  getOneGroupTimeSlotById = async (timeSlotId: string): Promise<TimeSlot> =>
+    this.timeSlotRepository.findOneBy({ _id: timeSlotId });
+
+  updateOneGroupTimeSlotById = async (
+    conditions: {
+      groupId: string;
+      timeSlotId: string;
+    },
```
Comment (MaxenceMottard) [inline src/services/time-slot.service.ts:38]:
```suggestion
    timeSlotId: string,
```

File: src/dto/patching-time-slot.dto.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsNotEmpty, IsString } from 'class-validator';
+import { TimeSlotDTO } from './time-slot.dto';
+
+export class PatchingTimeSlotDTO extends TimeSlotDTO {
+  @ApiProperty()
+  @IsNotEmpty()
+  @IsString()
+  groupdId: string;
```
Comment (MaxenceMottard) [inline src/dto/patching-time-slot.dto.ts:9]: I think you should not have `groupId`, but you should have `startDate` and `endDate`.

**Review by MaxenceMottard — CHANGES_REQUESTED** [review]

File: src/controllers/time-slot.controller.ts (hunk 8 above)
Comment (MrZyr0) [reply src/controllers/time-slot.controller.ts:64]: The DB request is made with a `groupId` not with a `organizationId `.

Screenshot 2022-02-20 at 14 49 49

File: src/services/time-slot.service.ts (hunk 11 above)
Comment (MrZyr0) [reply src/services/time-slot.service.ts:50]: Why ?

File: src/controllers/time-slot.controller.ts (hunk 8 above)
Comment (MaxenceMottard) [reply src/controllers/time-slot.controller.ts:63]: But actually you don't have `groupId` 😆

**Review by MaxenceMottard — APPROVED** [review]

**Pull request merged**